src/core/dense_set.cc

```diff
@@ -0,0 +1,558 @@
+    it->SetLink(lk);
+  }
+}
+
+void* DenseSet::PopFront(DenseSet::ChainVectorIterator it) {
+  if (IsEmpty(it)) {
+    return nullptr;
+  }
+
+  void* ret = it->GetObject();
+
+  // if this is the last object in this chain just reset the
+  // DensePtr instead of deleting it
+  if (it->IsObject()) {
+    it->Reset();
+  } else {
+    DCHECK(it->IsLink());
+    // set the current item in the list to the next one & deallocate the link data structure
+    DenseLinkKey* lk = (DenseLinkKey*)it->Raw();
+    *it = lk->next;
+    mr()->deallocate(lk, sizeof(DenseLinkKey), alignof(DenseLinkKey));
+  }
+
+  return ret;
+}
+
+DenseSet::DensePtr* DenseSet::Find(void* needle, DenseSet::ChainVectorIterator list,
+                                   DenseSet::DensePtr** prev_out) {
+  DensePtr* prev = nullptr;
+  ChainIterator it = ChainBegin(list);
+  for (; it != ChainEnd(list); ++it) {
+    if (Equal(*it, needle)) {
+      break;
+    }
+
+    prev = *it;
+  }
+
+  if (!it->IsEmpty() && prev_out != nullptr) {
+    *prev_out = prev;
+  }
+
+  return *it;
+}
+
+DenseSet::ChainIterator DenseSet::Unlink(DenseSet::ChainIterator node,
+                                          DenseSet::ChainIterator prev, DenseSet::DensePtr* out) {
+  DCHECK(out != nullptr);
+  if (node->IsObject()) {
+    // this is the only item in the list, reset it and keep the iterator here
+    if (prev == nullptr) {
+      // copy the pointer to the data before a reset
+      *out = **node;
+      node->Reset();
+      return node;
+    } else {
+      DCHECK(prev->IsLink());
+      /*
+        here we can try to pull a sneaky optimization, since the caller
+        may get a LinkKey or an regular object it is responsible for freeing it.
+        Some use cases (i.e Grow) may be able to use a LinkKey without freeing it as they can be
+        inserted into the front of a non-empty chained list. So instead of freeing them here
+        and save some memory management calls
+      */
+      DenseLinkKey* lk = (DenseLinkKey*)prev->Raw();
+      DCHECK(!lk->IsDisplaced());
+
+      // swap the two pointers to the actual data
+      void* lk_tmp = lk->GetObject();
+      lk->SetObject(node->GetObject());
+      node->SetObject(lk_tmp);
+
+      // prev now contains the data unlinked so store this pointer value in the out ptr
+      *out = **prev;
+      // overwrite prev with the data in node which is the end node containing the original
+      // previous value
+      **prev = **node;
+
+      // set the iterator to be the location of the previous value as this now contains
+      // the valid node still in the linked list
+      return prev;
+    }
+  } else {
+    DCHECK(node->IsLink());
+    // return the node and replace its spot with the next node in the list
+    DenseLinkKey* lk = (DenseLinkKey*)node->Raw();
+    *out = **node;
+    **node = lk->next;
+    return node;
+  }
+}
+
+DenseSet::ChainIterator DenseSet::UnlinkAndFree(DenseSet::ChainIterator node,
```

Comment:
You never use the returned `ChainIterator` but you return the value you use via `void**`. 
won't it be better to return `void*` as a result?

Reply:
yep. Same reasoning and change as the approach for `Unlink`